scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1493 @@
+)
+
+__all__ = ['scale', 'discrepancy', 'QMCEngine', 'Sobol', 'Halton',
+           'OrthogonalLatinHypercube', 'LatinHypercube', 'OptimalDesign',
+           'multinomial_qmc', 'NormalQMC', 'MultivariateNormalQMC']
+
+
+def scale(sample, bounds, reverse=False):
+    """Sample scaling from unit hypercube to bounds range.
+
+    To convert a sample from :math:`[0, 1)` to :math:`[a, b), b>a`, the
+    following transformation is used:
+
+    .. math::
+
+        (b - a) * sample + a
+
+    Parameters
+    ----------
+    sample : array_like (n_samples, dim)
+        Sample to scale.
+    bounds : tuple or array_like ([min, dim], [max, dim])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+    reverse : bool
+        Reverse the transformation, from bounds range to unit hypercube.
+
+    Returns
+    -------
+    sample : array_like (n_samples, dim)
+        Scaled-sample.
+
+    Examples
+    --------
+    >>> from scipy.stats import qmc
+    >>> bounds = [[-2, 0],
+    ...           [6, 5]]
+    >>> sample = [[0.5 , 0.5 ],
+    ...           [0.75, 0.25]]
+    >>> qmc.scale(sample, bounds)
+    array([[2.  , 2.5 ],
+           [4.  , 1.25]])
+
+    """
+    bounds = np.asarray(bounds)
+    min_ = np.min(bounds, axis=0)
+    max_ = np.max(bounds, axis=0)
+    if not reverse:
+        return sample * (max_ - min_) + min_
+    else:
+        return (sample - min_) / (max_ - min_)
+
+
+def discrepancy(sample, iterative=False, method='CD'):
+    """Discrepancy on a given sample.
+
+    Parameters
+    ----------
+    sample : array_like (n_samples, dim)
+        The sample to compute the discrepancy from.
+    iterative : bool
+        Must be False if not using it for updating the discrepancy.
+    method : str
+        Type of discrepancy, can be ['CD', 'WD', 'MD', 'star'].
+
+    Returns
+    -------
+    discrepancy : float
+        Centered discrepancy.
+
+    Notes
+    -----
+    The discrepancy is a uniformity criterion used to assess the space filling
+    of a number of samples in a hypercube.
+    The discrepancy measures how the spread of the points deviates from a
+    uniform distribution.
+    The lower the value is, the better the coverage of the parameter space is.
+
+    Taking a subspace of the parameter space we count the number of points in
+    the subpace and compare it to the total number of the points of the
+    sample. This value is substracted by the volume of the subspace. This
+    process is done over all subspaces and the highest value is kept.
+
+    Four methods are available:
+
+    * ``CD``: Centered Discrepancy - subspace involves a corner of the
+      hypercube
+    * ``WD``: Wrap-around Discrepancy - subspace can wrap around bounds
+    * ``MD``: Mixture Discrepancy - mix between CD/WD covering more criteria
```

Comment:
What does "covering more criteria" mean?

Reply:
Meaning more criteria to be called uniform. From Fang2006 you can see that there are a bunch of criteria to define all these metrics. All metrics try to give cover some specific metrics. Here are the criteria for the centered discrepancy for instance: 

[C1] It is invariant under permuting factors and/or runs.
[C2] It is invariant under rotation of the coordinates.
[C3] It can measure not only uniformity of P over Cs, but also the projection
uniformity of P over Cu, where u is a non-empty subset of {1, ··· , s}.
[C4] There is some reasonable geometric meaning.
[C5] It is easy to compute.
[C6] It satisfies the Koksma-Hlawka-like inequality.
[C7] It is consistent with other criteria in experimental design.

Reply:
Could you clarify this in the documentation?

Reply:
Thanks. Please leave these open so I can mark them resolved, as they help me keep track of what I was looking for.